actionview/lib/action_view/ripper_ast_parser.rb

```diff
@@ -0,0 +1,196 @@
+          self[0].type == :@ident &&
+          self[0][0] == name
+      end
+
+      def argument_nodes
+        raise unless fcall?
+        return [] if self[1].nil?
+        if self[1].last == false || self[1].last.type == :vcall
+          self[1][0...-1]
+        else
+          self[1][0..-1]
+        end
+      end
+
+      def string?
+        type == :string_literal
+      end
+
+      def variable_reference?
+        type == :var_ref
+      end
+
+      def vcall?
+        type == :vcall
+      end
+
+      def call?
+        type == :call
+      end
+
+      def variable_name
+        self[0][0]
+      end
+
+      def call_method_name
+        self.last.first
+      end
+
+      def to_string
+        raise unless string?
+        self[0][0][0]
+      end
+
+      def hash?
+        type == :bare_assoc_hash || type == :hash
+      end
+
+      def to_hash
+        if type == :bare_assoc_hash
+          hash_from_body(self[0])
+        elsif type == :hash && self[0] == nil
+          {}
+        elsif type == :hash && self[0].type == :assoclist_from_args
+          hash_from_body(self[0][0])
+        end
+      end
+
+      def hash_from_body(body)
+        body.map do |hash_node|
+          return nil if hash_node.type != :assoc_new
+
+          [hash_node[0], hash_node[1]]
+        end.to_h
+      end
+
+      def symbol?
+        type == :@label || type == :symbol_literal
+      end
+
+      def to_symbol
+        if type == :@label && self[0] =~ /\A(.+):\z/
+          $1.to_sym
+        elsif type == :symbol_literal && self[0].type == :symbol && self[0][0].type == :@ident
+          self[0][0][0].to_sym
+        else
+          raise "not a symbol?: #{self.inspect}"
+        end
+      end
+    end
+
+    class NodeParser < ::Ripper
+      PARSER_EVENTS.each do |event|
+        arity = PARSER_EVENT_TABLE[event]
+        if arity == 0 && event.to_s.end_with?("_new")
+          module_eval(<<-eof, __FILE__, __LINE__ + 1)
+            def on_#{event}(*args)
+              Node.new(:list, args, lineno: lineno(), column: column())
+            end
+          eof
+        elsif event.to_s.match?(/_add(_.+)?\z/)
+          module_eval(<<-eof, __FILE__, __LINE__ + 1)
+            begin; undef on_#{event}; rescue NameError; end
+            def on_#{event}(list, item)
+              list.push(item)
+              list
+            end
+          eof
+        else
+          module_eval(<<-eof, __FILE__, __LINE__ + 1)
+            begin; undef on_#{event}; rescue NameError; end
+            def on_#{event}(*args)
+              Node.new(:#{event}, args, lineno: lineno(), column: column())
+            end
+          eof
+        end
+      end
+
+      SCANNER_EVENTS.each do |event|
+        module_eval(<<-End, __FILE__, __LINE__ + 1)
+          def on_#{event}(tok)
+            Node.new(:@#{event}, [tok], lineno: lineno(), column: column())
+          end
+        End
+      end
+    end
+
+    class RenderCallParser < NodeParser
```

Comment:
Maybe we can rename this `RenderCallExtractor` to avoid confusing with similarly named `RenderParser`